rst_files/wald_friedman.rst

```diff
@@ -401,467 +345,201 @@ that appear on the right side of the Bellman equation as separate functions
-
-* store all of our parameters for us internally
-
-* represent the solution to our Bellman equation alongside the :math:`\alpha` and :math:`\beta` decision cutoffs
-
-* accompany many of the same functions used above which now act on the type directly
-
-* allow us, in addition, to simulate draws and the decision process under different prior beliefs
-
-
-.. code-block:: julia
-
-    #=
-
-    Author: Shunsuke Hori
-
-    =#
-
-    mutable struct WFSolution{TAV <: AbstractVector, TR<:Real}
-        J::TAV
-        lb::TR
-        ub::TR
-    end
-
-    struct WaldFriedman{TR <: Real,
-                        TI <: Integer,
-                        TAV1 <: AbstractVector,
-                        TAV2 <: AbstractVector}
-        c::TR
-        L0::TR
-        L1::TR
-        f0::TAV1
-        f1::TAV1
-        m::TI
-        pgrid::TAV2
-        sol::WFSolution
-    end
 
-    function WaldFriedman(c, L0, L1, f0, f1; m = 25)
-
-        pgrid = range(0.0, stop = 1.0, length = m)
-
-        # Renormalize distributions so nothing is "too" small
-        f0 = clamp.(f0, 1e-8, 1-1e-8)
-        f1 = clamp.(f1, 1e-8, 1-1e-8)
-        f0 = f0 / sum(f0)
-        f1 = f1 / sum(f1)
-        J = zeros(m)
-        lb = 0.
-        ub = 0.
-
-        WaldFriedman(c, L0, L1, f0, f1, m, pgrid, WFSolution(J, lb, ub))
-    end
-
-    current_distribution(wf::WaldFriedman, p::Real) = p * wf.f0 + (1 - p) * wf.f1
-
-    function bayes_update_k(wf, p, k)
-        f0_k = wf.f0[k]
-        f1_k = wf.f1[k]
-
-        p_tp1 = p * f0_k / (p * f0_k + (1 - p) * f1_k)
-
-        return clamp(p_tp1, 0, 1)
-    end
-
-    bayes_update_all(wf, p) = clamp.(p * wf.f0 ./ (p * wf.f0 + (1 - p) * wf.f1), 0, 1)
-
-    payoff_choose_f0(wf, p) = (1 - p) * wf.L0
-
-    payoff_choose_f1(wf, p) = p * wf.L1
-
-    function EJ(wf, p, J)
-        # Pull out information
-        f0, f1 = wf.f0, wf.f1
-
-        # Get the current believed distribution and tomorrows possible dists
-        # Need to clip to make sure things don't blow up (go to infinity)
-        curr_dist = current_distribution(wf, p)
-        tp1_dist = bayes_update_all(wf, p)
-
-        # Evaluate the expectation
-        EJ = dot(curr_dist, J.(tp1_dist))
-
-        return EJ
-    end
-
-    payoff_continue(wf, p, J) = wf.c + EJ(wf, p, J)
-
-    function bellman_operator(wf, J)
-        c, L0, L1, f0, f1 = wf.c, wf.L0, wf.L1, wf.f0, wf.f1
-        m, pgrid = wf.m, wf.pgrid
-
-        J_out = similar(J)
-        J_interp = LinearInterpolation(pgrid, J)
-
-        for (p_ind, p) in enumerate(pgrid)
-            # Payoff of choosing model 0
-            p_c_0 = payoff_choose_f0(wf, p)
-            p_c_1 = payoff_choose_f1(wf, p)
-            p_con = payoff_continue(wf, p, J_interp)
-
-            J_out[p_ind] = min(p_c_0, p_c_1, p_con)
+    function decision_rule(d0, d1, L0, L1, c)
+        function cost(p, d0, d1, L0, L1, c)
+            c ≥ zero(c) || throw(ArgumentError("Cost must be non-negative"))
```

Comment:
https://github.com/QuantEcon/lecture-source-jl/blob/master/style.md#general-control-structures-and-code-organization Avoid the short-circuiting pattern when you just want to do an assert.

These is very tough to read for an introductory user.